### PR TITLE
Ui 2020 improvements

### PR DESCRIPTION
**Pull request by Yvanovski**

### CODE REVIEW

**Comment by Yvanovski** [comment]
Closes #67

**Comment by Yvanovski** [comment]
Closes #85 , not committed at the moment but will be soon

**Comment by Yvanovski** [comment]
Closes #68 
Same as issue 85, I will just commit the current code right now so it will be clear that these issues are already fixed

**Comment by Yvanovski** [comment]
Closes #93

File: TrafficJam/App.config [hunk 1]
```diff
@@ -1,6 +1,38 @@
...
+      <dependentAssembly>
+        <assemblyIdentity name="System.Text.Encoding.CodePages" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-4.1.3.0" newVersion="4.1.3.0" />
+      </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="System.Memory" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-4.0.1.1" newVersion="4.0.1.1" />
+      </dependentAssembly>
+    </assemblyBinding>
+  </runtime>
```
Comment (KnapSac) [inline TrafficJam/App.config]: Please revert these changes, I am not sure what caused them, but they aren't supposed to be there.

File: TrafficJam/Forms/EntryPointsPanel.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        #region Fields
+
+        //private readonly Map _map = new Map( TrafficTracer, 1 );
+        //private readonly EntryPointsSub _subPanel = new EntryPointsSub( 0 );
+        //private readonly MainWindow mw = new MainWindow( trafficTracer:0 );
+        //MainWindow._trafficTracer
+
+        #endregion
+
+        public EntryPointsPanel()
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsPanel.cs]:
```suggestion
        public EntryPointsPanel( MainWindow mainWindow )
```

File: TrafficJam/Forms/EntryPointsPanel.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            Location = new Point( 50, 300 );
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsPanel.cs]:
```suggestion
            _mainWindow = mainWindow;
            Location = new Point( 50, 300 );
```

File: TrafficJam/Forms/EntryPointsPanel.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            Size = new Size( 500, 600 );
+            BorderStyle = BorderStyle.FixedSingle;
+
+            for ( int i = 0; i < 5 /*_mainWindow.TrafficTracer.Map.EntryPointsCount*/; i++ ) //TODO using test number, getting amount of entry points does not work yet
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsPanel.cs]:
```suggestion
            for ( int i = 0; i < _mainWindow.TrafficTracer.Map.EntryPointsCount; i++ )
```

File: TrafficJam/Forms/EntryPointsPanel.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+#endregion
+
+namespace TrafficJam.Forms
+{
+    internal class EntryPointsPanel : Panel
+    {
+        #region Fields
+
+        //private readonly Map _map = new Map( TrafficTracer, 1 );
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsPanel.cs]:
```suggestion

```

File: TrafficJam/Forms/EntryPointsPanel.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        //private readonly EntryPointsSub _subPanel = new EntryPointsSub( 0 );
+        //private readonly MainWindow mw = new MainWindow( trafficTracer:0 );
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsPanel.cs]:
```suggestion
        private readonly MainWindow _mainWindow;
```

File: TrafficJam/Forms/EntryPointsPanel.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        //MainWindow._trafficTracer
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsPanel.cs]:
```suggestion

```

File: TrafficJam/Forms/EntryPointsSub.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    internal class EntryPointsSub : Panel
+    {
+        #region Fields
+
+        private readonly Font _trebuchetSmall = new Font( "Trebuchet MS", 8f );
+        private readonly Font _trebuchetLarge = new Font( "Trebuchet MS", 11f );
+
+        #endregion
+
+        public EntryPointsSub( int y )
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsSub.cs]:
```suggestion
        public EntryPointsSub( EntryPointsPanel entryPointsPanel, int y )
```

This subpanel also needs to be aware of his owner, so they can report back to the application when a value changes.

File: TrafficJam/Forms/EntryPointsSub.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+using System.Windows.Forms;
+
+#endregion
+
+namespace TrafficJam.Forms
+{
+    internal class EntryPointsSub : Panel
+    {
+        #region Fields
+
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsSub.cs]:
```suggestion

            private readonly EntryPointsPanel _entryPointsPanel;
```

File: TrafficJam/Forms/EntryPointsSub.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        private readonly Font _trebuchetSmall = new Font( "Trebuchet MS", 8f );
+        private readonly Font _trebuchetLarge = new Font( "Trebuchet MS", 11f );
+
+        #endregion
+
+        public EntryPointsSub( int y )
+        {
+            Point basePoint = new Point(0, y * 100);
```
Comment (KnapSac) [inline TrafficJam/Forms/EntryPointsSub.cs]:
```suggestion
            _entryPointsPanel = entryPointsPanel;
            Point basePoint = new Point(0, y * 100);
```

File: TrafficJam/Forms/StartupWindow.cs [hunk 11]
```diff
@@ -37,15 +37,16 @@ public StartupWindow( TrafficTracer trafficTracer )
...
             _situation1 = _resourceManager.GetObject( "Map1" ) as Image;
             _situation2 = _resourceManager.GetObject( "Map1" ) as Image;
             _situation3 = _resourceManager.GetObject( "Map1" ) as Image;
 
             _mapPreview = new PictureBox( )
             {
-                Location = new Point( 725, 340 ),
+                Location = new Point( 725, 308 ),
                 Image = _situation1,
+                Size = new Size(100,100),
```
Comment (KnapSac) [inline TrafficJam/Forms/StartupWindow.cs]:
```suggestion
               
```
There is no need to set the ```Size```, because that is calculated automatically because ```AutoSize``` is set to ```true```

File: TrafficJam/Forms/StartupWindow.cs [hunk 12]
```diff
@@ -37,15 +37,16 @@ public StartupWindow( TrafficTracer trafficTracer )
 
             _trafficTracer = trafficTracer;
 
-            //TODO Added different Map layouts for the other Maps
+            //TODO Add different Map layouts for the other Maps
```
Comment (KnapSac) [inline TrafficJam/Forms/StartupWindow.cs]:
```suggestion
            
```
This is already done in develop

File: TrafficJam/Forms/ToolsPanel.cs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        }
+
+        private void CrossRoad_Click(object sender, EventArgs e)
+        {
+            _selectedToolButton = _crossRoad;
+            BorderSizeToolMenu();
+
+            CurrentTool = new CrossRoad();
+            Console.WriteLine("tool is crossroad");
```
Comment (KnapSac) [inline TrafficJam/Forms/ToolsPanel.cs]:
```suggestion  
```
We don't need this output

File: TrafficJam/Forms/ToolsPanel.cs [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        private void RoadBlock_Click(object sender, EventArgs e)
+        {
+            _selectedToolButton = _roadBlock;
+            BorderSizeToolMenu();
+
+            CurrentTool = new RoadBlock();
+            Console.WriteLine("tool is roadblock");
```
Comment (KnapSac) [inline TrafficJam/Forms/ToolsPanel.cs]:
```suggestion
```

File: TrafficJam/Forms/ToolsPanel.cs [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        private void SpeedBump_Click(object sender, EventArgs e)
+        {
+            _selectedToolButton = _speedBump;
+            BorderSizeToolMenu();
+
+            CurrentTool = new SpeedBump();
+            Console.WriteLine("tool is speedbump");
```
Comment (KnapSac) [inline TrafficJam/Forms/ToolsPanel.cs]:
```suggestion
```

File: TrafficJam/Forms/ToolsPanel.cs [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        private void SpeedSign_Click(object sender, EventArgs e)
+        {
+            _selectedToolButton = _speedSign;
+            BorderSizeToolMenu();
+
+            CurrentTool = new SpeedSign();
+            Console.WriteLine("tool is speedsign");
```
Comment (KnapSac) [inline TrafficJam/Forms/ToolsPanel.cs]:
```suggestion
```

File: TrafficJam/Forms/ToolsPanel.cs [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        private void TrafficLight_Click(object sender, EventArgs e)
+        {
+            _selectedToolButton = _trafficLight;
+            BorderSizeToolMenu();
+
+            CurrentTool = new TrafficLight();
+            Console.WriteLine("tool is traffic light");
```
Comment (KnapSac) [inline TrafficJam/Forms/ToolsPanel.cs]:
```suggestion
```

File: TrafficJam/Forms/ToolsPanel.cs [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        private void throughputMeter_Click(object sender, EventArgs e)
+        {
+            _selectedToolButton = _throughputMeter;
+            BorderSizeToolMenu();
+
+            CurrentTool = new ThroughputMeter();
+            Console.WriteLine("tool is throughput meter");
```
Comment (KnapSac) [inline TrafficJam/Forms/ToolsPanel.cs]:
```suggestion
```

File: TrafficJam/TrafficJam.csproj [hunk 19]
```diff
@@ -161,30 +188,48 @@
...
+    </Compile>
+    <Compile Include="Forms\EntryPointsSub.cs">
+      <SubType>Component</SubType>
+    </Compile>
+    <Compile Include="Forms\ToolsPanel.cs">
+      <SubType>Component</SubType>
+    </Compile>
+    <Compile Include="Forms\TrafficFlowPanel.cs">
+      <SubType>Component</SubType>
+    </Compile>
```
Comment (KnapSac) [inline TrafficJam/TrafficJam.csproj]: Keep these Compile includes for the panels, but revert the rest. Same story as with the ```app.config``` file, these are unwanted changes.

File: TrafficJam/TrafficJam.csproj.DotSettings [hunk 20]
```diff
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (KnapSac) [inline TrafficJam/TrafficJam.csproj.DotSettings]: You should save this to your personal settings file, not to a file which gets committed. That way, ReSharper stores the setting for all your projects, not only this one. After you have saved it to your personal settings file, please remove this file.

File: TrafficJam/packages.config [hunk 21]
```diff
@@ -21,30 +26,36 @@
...
-  <package id="System.Text.RegularExpressions" version="4.3.0" targetFramework="net48" />
+  <package id="System.Text.RegularExpressions" version="4.3.1" targetFramework="net48" />
   <package id="System.Threading" version="4.3.0" targetFramework="net48" />
   <package id="System.Threading.Tasks" version="4.3.0" targetFramework="net48" />
+  <package id="System.Threading.Tasks.Extensions" version="4.5.3" targetFramework="net48" />
   <package id="System.Threading.Timer" version="4.3.0" targetFramework="net48" />
-  <package id="System.Xml.ReaderWriter" version="4.3.0" targetFramework="net48" />
+  <package id="System.Xml.ReaderWriter" version="4.3.1" targetFramework="net48" />
   <package id="System.Xml.XDocument" version="4.3.0" targetFramework="net48" />
 </packages>
```
Comment (KnapSac) [inline TrafficJam/packages.config]: Same thing as with ```app.config```, please revert.
